pyctcdecode/alphabet.py

```
@@ -1,91 +1,111 @@
+    is_bpe = any([s.startswith("##") for s in labels]) or any(
+        [s.startswith(BPE_TOKEN) for s in labels]
+    )
+    if is_bpe:
+        logger.info("Alphabet determined to be of BPE style.")
     else:
-        clean_labels[ctc_token_idx] = ""
-    return clean_labels
+        logger.info("Alphabet determined to be of regular style.")
+    return is_bpe
+
+
+def _normalize_regular_alphabet(labels: List[str]) -> List[str]:
+    """Normalize non-bpe labels to alphabet for decoder."""
+    normalized_labels = labels[:]
+    # substitute space characters
+    if "|" in normalized_labels and " " not in normalized_labels:
+        logger.info("Found '|' in vocabulary but not ' ', doing substitution.")
+        normalized_labels[normalized_labels.index("|")] = " "
+    # substituted ctc blank char
+    for n, label in enumerate(normalized_labels):
+        if BLANK_TOKEN_PTN.match(label):
+            logger.info("Found %s in vocabulary, substituting with %s.", label, "")
+            normalized_labels[n] = ""
+    if "_" in normalized_labels and "" not in normalized_labels:
+        logger.info("Found '_' in vocabulary but not '', doing substitution.")
+        normalized_labels[normalized_labels.index("_")] = ""
+    if "" not in normalized_labels:
+        logger.info("CTC blank char '' not found, appending to end.")
+        normalized_labels.append("")
+    # substitute unk
+    for n, label in enumerate(normalized_labels):
+        if UNK_TOKEN_PTN.match(label):
+            logger.info("Found %s in vocabulary, substituting with %s.", label, UNK_TOKEN)
+            normalized_labels[n] = UNK_TOKEN
+    # additional checks
+    if any([len(c) > 1 for c in normalized_labels]):
+        logger.warning(
+            "Found entries of length > 1 in alphabet. This is unusual unless style is BPE. "
+            "Is this correct?"
+        )
+    if " " not in normalized_labels:
+        logger.warning("Space token ' ' missing from vocabulary.")
```

not to overthink this too much, but maybe worth considering adding a `strict` param that, when set, will raise instead of logging here.

---

i'd say worth it if we check with for n=2 or even n=3 characters, no?